### PR TITLE
Add set price contract, refactor buy contract

### PR DESCRIPTION
**Pull request by samuelWilliams99**
This ticket is acting in preparation for the gifting contract, which will submit 2 transactions, a set price then a buy.
This also allows for greater code reuse when running transactions against seabug _not_ on the marketplace, which the gifting transaction will be.
Lastly, this changes the buying logic to now return the bought token to the user, rather than immediated putting it back up for sale on the market. There is a simple sum type for defining this behaviour, if you need to change it for debugging.

### CODE REVIEW

**Review by rynoV — COMMENTED** [review]
Buying seems to work after a small change (and minting a new nft using the new minting policy): https://testnet.cardanoscan.io/transaction/910f14c0d53e7466f5ba7ac2e3d20a51f30344ba74808aa8ebeadf143ee7a885

I'm sure the buyers would've liked if that contract worked ;)

**Comment by samuelWilliams99** [comment]
Ah neat, thanks! Would you mind running the formatter?

File: src/Seabug/Contract/Buy.purs [hunk 1]
```diff
@@ -45,7 +45,9 @@ mkBuyTxData nftData mScriptUtxos = do
     nftData
     mScriptUtxos
   let
-    nftCollection = unwrap (unwrap nftData).nftCollection
+    nftData' = unwrap nftData
+    nftCollection = unwrap nftData'.nftCollection
+    nftId = unwrap nftData'.nftId
```
Comment (samuelWilliams99) [inline src/Seabug/Contract/Buy.purs:50]: Given these, the nftPrice line below can probably be changed to just grab from `nftId`

**Review by samuelWilliams99 — COMMENTED** [review]
I'm done for today, but if you could make this change then review at some point today, that'd be great :)

**Comment by rynoV** [comment]
For some reason in at least one case the buyer isn't covering the full price of the nft, and it looks like marketplace script's ada is being used to cover part of it. This can be seen in the previous tx I linked (unless I'm reading wrong). Still checking if this was introduced by these changes

**Comment by samuelWilliams99** [comment]
Oh that makes sense, this was always possible, its just that returning the utxo to the market place meant you still had to cover that min ada - we'll have to discuss whether this should be addressed, and if so - how.

**Comment by samuelWilliams99** [comment]
See [here](https://github.com/mlabs-haskell/seabug/issues/28), we'll address this later. @rynoV please do review when you get a chance :)

File: src/Seabug/Contract/Util.purs [hunk 2]
```diff
@@ -1,20 +1,209 @@
...
+    newNftValue = Value.singleton curr newName one
+    mintRedeemer = Redeemer $ toData $ act nftData.nftId
+
+    containsNft :: forall (a :: Type). (a /\ TransactionOutput) -> Boolean
+    containsNft (_ /\ TransactionOutput out) =
+      Value.valueOf out.amount curr oldName == one
+
+  userAddr <- liftedM (name <> ": Cannot get user addr") getWalletAddress
+  userUtxos <-
+    liftedM (name <> ": Cannot get user Utxos") $ utxosAt userAddr
```
Comment (rynoV) [inline src/Seabug/Contract/Util.purs:182]: I'm pretty sure these two queries are unnecessary when `mScriptUtxos` is not `Nothing`. The balancer will include the utxos at `getWalletAddress` anyways (and make the query anyways). Could refactor to make this query only when `mScriptUtxos` is `Nothing`, but might not be a priority to implement and test that if we're trying to get this merged quickly

**Review by rynoV — APPROVED** [review]
Lgtm for the most part. I double checked buying is working: https://testnet.cardanoscan.io/transaction/3a1edb857a2e3ba1bd2eaa86de505ec19252e28dda965f94048fd58509c32c05

File: src/Seabug/Contract/Util.purs (hunk 2 above)
Comment (samuelWilliams99) [reply src/Seabug/Contract/Util.purs:182]: Oh good point, I'll have a look

Comment (rynoV) [reply src/Seabug/Contract/Util.purs:182]: An untested suggestion for a starting point @samuelWilliams99 :

```diff
modified   src/Seabug/Contract/Util.purs
@@ -51,7 +51,7 @@ import Data.Array (find) as Array
 import Data.Bifunctor (lmap)
 import Data.BigInt (BigInt)
 import Data.BigInt as BigInt
-import Data.Map (insert, toUnfoldable)
+import Data.Map (singleton, toUnfoldable) as Map
 import Plutus.Types.Transaction (UtxoM)
 import Seabug.MarketPlace (marketplaceValidator)
 import Seabug.Metadata.Share (mkShare)
@@ -177,20 +177,21 @@ mkChangeNftIdTxData name act mapNft (NftData nftData) mScriptUtxos = do
     containsNft (_ /\ TransactionOutput out) =
       Value.valueOf out.amount curr oldName == one
 
-  userAddr <- liftedM (name <> ": Cannot get user addr") getWalletAddress
-  userUtxos <-
-    liftedM (name <> ": Cannot get user Utxos") $ utxosAt userAddr
-  utxo /\ utxoIndex <-
-    liftContractM (name <> ": NFT not found")
-      $ Array.find containsNft
-      $ toUnfoldable
-      $ unwrap
-      $ fromMaybe userUtxos mScriptUtxos
+    findNftUtxo = liftContractM (name <> ": NFT not found")
+      <<< Array.find containsNft
+      <<< Map.toUnfoldable
+      <<< unwrap
+
+  utxo /\ utxoIndex <- findNftUtxo =<< case mScriptUtxos of
+    Nothing -> do
+      userAddr <- liftedM (name <> ": Cannot get user addr") getWalletAddress
+      liftedM (name <> ": Cannot get user Utxos") $ utxosAt userAddr
+    Just scriptUtxos -> pure scriptUtxos
+
   let
-    utxosForTx = insert utxo utxoIndex $ unwrap userUtxos
     lookups = mconcat
       [ ScriptLookups.mintingPolicy policy
-      , ScriptLookups.unspentOutputs utxosForTx
+      , ScriptLookups.unspentOutputs $ Map.singleton utxo utxoIndex
       ]
 
     constraints = mustMintValueWithRedeemer mintRedeemer
```

Comment (rynoV) [reply src/Seabug/Contract/Util.purs:182]: @samuelWilliams99 I could just pick this up if your testing environment isn't set up yet, might be quicker

Comment (samuelWilliams99) [reply src/Seabug/Contract/Util.purs:182]: Oh yeah actually, that'd be great thanks!

Comment (rynoV) [reply src/Seabug/Contract/Util.purs]: Implemented and tested: https://testnet.cardanoscan.io/transaction/4e306f28cde044f8c091cb808218c5b6d97d6f3c6b655d511dc4231ce5736f12

Also checked that the extra queries aren't being made

**Review by rynoV — APPROVED** [review]
Should be good to go now.

The set price contract isn't tested, but it's not used and most of it's code is shared by the buy contract which was tested

**Pull request merged**